File: service-database/src/main/java/com/bnorm/barkeep/db/DbBarService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db;

import java.util.List;

import javax.persistence.EntityManager;

import com.bnorm.barkeep.model.Bar;
import com.bnorm.barkeep.model.Ingredient;
import com.bnorm.barkeep.service.BarService;

public class DbBarService extends AbstractDbService implements BarService {

  public DbBarService(EntityManager entityManager) {
    super(entityManager);
  }

  @Override
  public List<Bar> getBars() {
    return super.getBars();
  }

  @Override
  public BarEntity getBar(long barId) {
    return super.getBar(barId);
  }

  @Override
  public BarEntity createBar(Bar bar) {
    if (bar.getId() != null) {
      throw new IllegalArgumentException(String.format("Cannot create bar that already has an id=%d", bar.getId()));
    } else if (bar.getOwner() == null) {
      throw new IllegalArgumentException("Cannot create bar without an owner");
    }
    UserEntity userEntity = find(bar.getOwner());
    if (userEntity == null) {
      throw new IllegalArgumentException(String.format("Cannot create bar with an unknown owner id=%d",
                                                       bar.getOwner().getId()));
    }

    return transaction(em -> {
      BarEntity barEntity = new BarEntity();

      barEntity.setTitle(bar.getTitle());
      barEntity.setDescription(bar.getDescription());
      barEntity.setOwner(userEntity);
      barEntity.addUser(userEntity);
      if (bar.getIngredients() != null) {
        for (Ingredient ingredient : bar.getIngredients()) {
          barEntity.addIngredient(find(ingredient));
        }
      }

      em.persist(barEntity);

      userEntity.addBar(barEntity);
      return barEntity;
    });
  }

  @Override
  public BarEntity setBar(long barId, Bar bar) {
    BarEntity barEntity = findBar(barId);
    requireExists(barEntity, barId, "bar");
    requireMatch(bar, barId, "bar");

    return transaction(em -> {
      if (bar.getTitle() != null) {
        barEntity.setTitle(bar.getTitle());
      }
      if (bar.getDescription() != null) {
        barEntity.setDescription(bar.getDescription());
      }
      if (bar.getOwner() != null) {
        barEntity.setOwner(find(bar.getOwner()));
      }
      if (bar.getIngredients() != null) {
        // todo is this correct?
        for (Ingredient ingredient : bar.getIngredients()) {
          barEntity.addIngredient(find(ingredient));
        }
      }

      return barEntity;
    });
  }

  @Override
  public void deleteBar(long barId) {
    transaction(em -> {
      BarEntity barEntity = findBar(barId);
      requireExists(barEntity, barId, "bar");
      em.remove(barEntity);
    });
  }
}

File: service-database/src/main/kotlin/com/bnorm/barkeep/db/DbBarService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db

import com.bnorm.barkeep.model.Bar
import com.bnorm.barkeep.model.BarSpec
import com.bnorm.barkeep.service.BarService
import javax.persistence.EntityManager

class DbBarService(private val em: EntityManager,
                   private val ingredientService: DbIngredientService,
                   private val userService: DbUserService) : BarService {

  override fun getBars(): List<Bar> {
    return em.createNamedQuery("BarEntity.findAll", BarEntity::class.java).resultList
  }

  fun findBar(id: Long): BarEntity? {
    return em.find(BarEntity::class.java, id)
  }

  override fun getBar(id: Long): BarEntity {
    return findBar(id) ?: throw IllegalArgumentException("Cannot find bar with id=$id")
  }

  override fun createBar(bar: BarSpec): BarEntity {
    val owner = bar.owner ?: throw IllegalArgumentException("Cannot create bar without an owner")
    val userEntity = userService.findUser(owner.id) ?: throw IllegalArgumentException("Cannot create bar with an unknown owner id=${owner.id}")

    val barEntity = BarEntity()
    barEntity.title = bar.title
    barEntity.description = bar.description
    barEntity.owner = userEntity
    barEntity.addUser(userEntity)

    em.txn {
      em.persist(barEntity)
      userEntity.addBar(barEntity)
    }
    return barEntity
  }

  override fun addBarIngredient(barId: Long, ingredientId: Long) {
    val barEntity = getBar(barId)
    val ingredientEntity = ingredientService.getIngredient(ingredientId)

    em.txn {
      barEntity.addIngredient(ingredientEntity)
    }
  }

  override fun removeBarIngredient(barId: Long, ingredientId: Long) {
    val barEntity = getBar(barId)
    val ingredientEntity = ingredientService.getIngredient(ingredientId)

    em.txn {
      barEntity.removeIngredient(ingredientEntity)
    }
  }

  override fun setBar(id: Long, bar: BarSpec): BarEntity {
    val barEntity = getBar(id)

    em.txn {
      // todo what if we want to clear a value?
      bar.title?.apply { barEntity.title = this }
      bar.description?.apply { barEntity.description = this }
      bar.owner?.apply { barEntity.owner = userService.getUser(this.id) }
    }
    return barEntity
  }

  override fun deleteBar(id: Long) {
    em.txn {
      em.remove(getBar(id))
    }
  }
}
